src/pw/fpga/fft_fpga.c

```diff
@@ -0,0 +1,353 @@
+ *  CP2K: A general program to perform molecular dynamics simulations        *
+ *  Copyright (C) 2000 - 2019  CP2K developers group                         *
+ *****************************************************************************/
+
+/******************************************************************************
+ *  Author: Arjun Ramaswami
+ *****************************************************************************/
+
+#if defined ( __PW_FPGA )
+
+// global dependencies
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdbool.h>
+
+// common dependencies
+#include "CL/opencl.h"
+
+// local dependencies
+#include "fft_fpga.h"
+#include "opencl_utils.h"
+
+// host variables
+static cl_platform_id platform = NULL;
+static cl_device_id device = NULL;
+static cl_context context = NULL;
+static cl_program program = NULL;
+
+static cl_command_queue queue1 = NULL, queue2 = NULL, queue3 = NULL;
+static cl_command_queue queue4 = NULL, queue5 = NULL, queue6 = NULL;
+
+static cl_kernel fft_kernel = NULL, fft_kernel_2 = NULL;
+static cl_kernel fetch_kernel = NULL, transpose_kernel = NULL, transpose_kernel_2 = NULL;
+
+// Device memory buffers
+cl_mem d_inData, d_outData;
+
+// Global Variables
+static cl_int status = 0;
+static int fft_size[3] = {0,0,0};
+bool fft_size_changed = true;
+
+// Function prototypes
+bool init();
+void cleanup();
+void init_program(char *data_path);
+void queue_setup();
+void queue_cleanup();
+void fftfpga_run_3d(bool inverse, cmplx *c_in);
+
+// --- CODE -------------------------------------------------------------------
+
+int pw_fpga_initialize_(){
+   status = init();
+   return status;
+}
+
+void pw_fpga_final_(){
+   cleanup();
+}
+
+/******************************************************************************
+ * \brief  check whether FFT3d can be computed on the FPGA or not. This depends 
+ *         on the availability of bitstreams whose sizes are for now listed here 
+ *         The var fft_size_changed indicates whether a new binary needs to be 
+ *         loaded before execution. 
+ * \param  N - integer pointer to the size of the FFT3d
+ * \retval true if fft3d size supported
+ *****************************************************************************/
+bool pw_fpga_check_bitstream_(int N[3]){
+    // check the supported sizes
+    if( (N[0] == 16 && N[1] == 16 && N[2] == 16) ||
+        (N[0] == 32 && N[1] == 32 && N[2] == 32) ||
+        (N[0] == 64 && N[1] == 64 && N[2] == 64)  ){
+
+        // if previously used binary needs to be executed again, otherwise set
+        // the new size and load new binary
+        if( fft_size[0] == N[0] && fft_size[1] == N[1] && fft_size[2] == N[2] ){
+            fft_size_changed = false;
+        }
+        else{
+            fft_size[0] = N[0];
+            fft_size[1] = N[1];
+            fft_size[2] = N[2];
+            fft_size_changed = true;
+        }
+        return 1;
+    }
+    else{
+        return 0;
```

Comment:
Sorry, I don't understand what you mean here.

Reply:
Nevermind, with the `fft_size` declared here it is clearer what's going on here.

But, what happens if I call this function twice with different but valid `data_path` and/or `N`s without `deinit` in-between? Are we leaking memory if things like `fft_kernel` get simply overwritten?

And what happens if I call it like

```
pw_fpga_check_bitstream_("/tmp/my.dat", {16,16,16});
cleanup();
pw_fpga_check_bitstream_("/tmp/my.dat", {16,16,16}); // due to state tracking here for 'caching' it will properly reinitialize the structures as expected I guess
```

Reply:
I see, successive calls with different `N`s causes memory leaks since it does another `init` without cleaning up. Will modify that (done!). 
Calling the function twice with different datapaths having the same `N` would end up using the first `N`'s program that was already loaded. Currently, there are fixed paths to find the binaries, this could be changed in the future to use env variable/ input file/ command line args. 
The code snippet example will cause failure because the binary / program will not be available to create buffers.